### PR TITLE
Cylc Wrapper uses arbitrary env set at CYLC_ENV_NAME

### PR DESCRIPTION
**Pull request by wxtim**
These changes address the bulk of #4165 

**Ticket Specific Checklist**
- [x] Use a new variable, `CYLC_ENV_NAME`, to allow fully arbitrary names.
- [ ] Also consider CYLC_ENVIRONMENT_TYPE. Not considered. Mind if I leave it to a future PR?
- [ ] Testing?
  - [x] Manual testing
    - [x] `CYLC_ENV_NAME=foo cylc version --long`
    - [x] `CYLC_ENV_NAME=foo CYLC_VERSION=8.0b3 cylc version --long`
    - [x] `CYLC_VERSION=8.0b3 cylc version --long`
    - [x] Ran a workflow with a remote task and checked the logfiles
  - [ ] Automated testing @hjoliver  - any thoughts? Are there any existing automated tests for this functionality? 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included.

### CODE REVIEW

File: usr/bin/cylc
Comment (oliver-sanders) [inline usr/bin/cylc:162]: We would expect *both* the env name and the version to be specified much of the time so these two need to be exclusive.

The env name should be used in preference to the version.

**Comment by oliver-sanders** [comment]
> CYLC_ENVIRONMENT_TYPE

Actually hold off on that, I think it might become obsolete with https://github.com/cylc/cylc-flow/issues/3704

> Testing? 

Manual for now.

> Automated testing

Manual for now.

File: usr/bin/cylc
Comment (wxtim) [reply usr/bin/cylc]: > Note we need to export the env name in this script.

Why?

**Comment by wxtim** [comment]
> Manual for now.

😢 😞 😿

File: usr/bin/cylc
Comment (wxtim) [reply usr/bin/cylc]: > so these two need to be exclusive.

I thought that I had made them exclusive:
```
if CYLC_VERSION
  use it
if CYLC_ENV_NAME
  use it (regardless of whether you just used it for CYLC_VERSION)
else:
  whateva
```
But I think I've made the logic clearer.

File: usr/bin/cylc
Comment (hjoliver) [inline usr/bin/cylc:148]:
```suggestion
# CYLC_ENV_NAME construction:
```

File: usr/bin/cylc
Comment (hjoliver) [inline usr/bin/cylc:157]: Nice :grin:

Comment (hjoliver) [reply usr/bin/cylc:157]: Note the comments from lines 23-114 will need updating.  (Possibly to be moved to the user guide in due course).

File: usr/bin/cylc
Comment (hjoliver) [inline usr/bin/cylc:168]: I don't think we need this `export`?  I guess it would allow a tiny shortcut in the scripting above if `cylc` is invoked again in subshells, but it would be clearer to have subshells work out the path in exactly the same way as the parent.

Ping @oliver-sanders as it was evidently his suggestion?

**Comment by hjoliver** [comment]
> Automated testing @hjoliver - any thoughts? Are there any existing automated tests for this functionality?

There's been no automated wrapper testing to date.  I'm sure we could do it by setting up the right environment in a container, but might be best to leave that to 8.x.

File: usr/bin/cylc
Comment (oliver-sanders) [reply usr/bin/cylc:186]: We may have multiple environments containing the same Cylc version which could lead to different results when activating via the `CYLC_VERSION` or `CYLC_ENV_NAME`.

File: usr/bin/cylc
Comment (wxtim) [reply usr/bin/cylc:175]: I've had a look.

File: usr/bin/cylc
Comment (hjoliver) [reply usr/bin/cylc:186]: I don't think we even need to set it here. By this point, it is used to set CYLC_HOME, and that's what's used to activate the environment.

**Comment by hjoliver** [comment]
(Seems to be working well, in my testing)

File: usr/bin/cylc
Comment (dpmatthews) [reply usr/bin/cylc:186]: I think the only case where we need to `export CYLC_ENV_NAME` is when we're using the default symlink version. We have to export it because cylc will always set `CYLC_VERSION` and we can no longer assume that the symlink version points to `cylc-$CYLC_VERSION`.

File: usr/bin/cylc
Comment (wxtim) [inline usr/bin/cylc:176]:
```suggestion
            # does not get overridden by CYLC_VERSION
```

**Review by dpmatthews — APPROVED** [review]
Not tested by me but logic looks good

**Comment by wxtim** [comment]
@hjoliver  & @oliver-sanders , Dave has now approved the logic. Might you test the functionality?

**Comment by hjoliver** [comment]
> @hjoliver & @oliver-sanders , Dave has now approved the logic. Might you test the functionality?

:+1: I'll take a final look and test first thing tomorrow.

File: usr/bin/cylc
Comment (hjoliver) [inline usr/bin/cylc:67]:
```suggestion
# Alternatively use CYLC_ENV_NAME to use Cylc from an arbitrary Conda or Python
# virtual environment:
```

File: usr/bin/cylc
Comment (hjoliver) [inline usr/bin/cylc:74]:
```suggestion
# (or similarly for a native Python venv)
```

**Review by hjoliver — APPROVED** [review]
Looks good, thanks @wxtim

**Comment by hjoliver** [comment]
Worth noting in the comments (and later, user guide) that `CYLC_ENV_NAME` can be an absolute path to an environment, or relative to `CYLC_HOME_ROOT` - particularly as the latter can be needed when setting up a matching venv on a remote job host. (Tested).  This is only a comment however, so I'll merge this now and add it along with a change log entry.

File: usr/bin/cylc
Comment (hjoliver) [reply usr/bin/cylc]: (Oops redundant with subsequent comments, I'll remove in the change log commit)

**Pull request merged**
